File: typescript/tests/gpu-catalog-integrity.test.ts
/**
 * GPU catalog integrity — structure, Decimal parsing, freshness, SKU consistency.
 *
 * Ports python/tests/test_gpu_catalog_integrity.py to vitest. Pins the
 * STRUCTURAL invariants so a future refresh can't drift shape; freshness
 * check enforces Decision #11's tighter 90/365-day thresholds.
 */

import { describe, expect, test } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import Decimal from "decimal.js";

const HERE = dirname(fileURLToPath(import.meta.url));
const CATALOG_PATH = join(HERE, "../src/data/gpu_prices.json");

function load(): Record<string, any> {
  return JSON.parse(readFileSync(CATALOG_PATH, "utf-8")) as Record<string, any>;
}

function isoDate(s: string): Date {
  // Strict ISO-8601 YYYY-MM-DD parser — throws on malformed input.
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s)) {
    throw new Error(`not ISO-8601 date: ${s}`);
  }
  const d = new Date(`${s}T00:00:00Z`);
  if (Number.isNaN(d.getTime())) throw new Error(`not valid date: ${s}`);
  return d;
}

function daysBetween(a: Date, b: Date): number {
  return Math.round((a.getTime() - b.getTime()) / (1000 * 60 * 60 * 24));
}

describe("GPU catalog integrity", () => {
  test("catalog parses as JSON with _meta", () => {
    const data = load();
    expect(data).toHaveProperty("_meta");
  });

  test("_meta has required default keys (all 4 billing models)", () => {
    const data = load();
    const meta = data._meta as Record<string, any>;
    const required = [
      "version",
      "last_updated",
      "currency",
      "default_per_instance_hour_usd",
      "default_per_gpu_second_active_usd",
      "default_per_gpu_hour_reserved_usd",
      "default_per_vgpu_hour_usd",
      "description",
      "notes",
    ];
    for (const k of required) {
      expect(meta).toHaveProperty(k);
      if (k.startsWith("default_") && k.endsWith("_usd")) {
        // Decimal-parseable.
        new Decimal(String(meta[k]));
      }
    }
    expect(meta.currency).toBe("USD");
  });

  test("all 8 providers present", () => {
    const data = load();
    const expected = new Set([
      "aws",
      "gcp",
      "azure",
      "modal",
      "runpod",
      "lambda_labs",
      "coreweave",
      "replicate",
    ]);
    for (const e of expected) {
      expect(data).toHaveProperty(e);
    }
  });

  test("every provider has _last_verified as ISO-8601 date", () => {
    const data = load();
    for (const [k, block] of Object.entries(data)) {
      if (k === "_meta") continue;
      const lv = (block as Record<string, any>)._last_verified as string;
      isoDate(lv); // throws on bad
    }
  });

  test("Decision #11 soft-warn at 90 days (deterministic simulation)", () => {
    const data = load();
    const dates = Object.entries(data)
      .filter(([k]) => k !== "_meta")
      .map(([, b]) => isoDate((b as Record<string, any>)._last_verified));
    const earliest = dates.reduce(
      (acc, d) => (d < acc ? d : acc),
      dates[0],
    );
    const fakeToday = new Date(
      earliest.getTime() + 91 * 24 * 60 * 60 * 1000,
    );
    const warnings: string[] = [];
    for (const [provider, block] of Object.entries(data)) {
      if (provider === "_meta") continue;
      const verified = isoDate(
        (block as Record<string, any>)._last_verified,
      );
      if (daysBetween(fakeToday, verified) > 90) {
        warnings.push(
          `gpu_prices.json: ${provider} _last_verified is ${daysBetween(fakeToday, verified)} days old (soft limit 90)`,
        );
      }
    }
    expect(warnings.some((w) => w.includes("soft limit 90"))).toBe(true);
  });

  test("Decision #11 hard-fail at 365 days", () => {
    const data = load();
    const today = new Date();
    const stale: string[] = [];
    for (const [provider, block] of Object.entries(data)) {
      if (provider === "_meta") continue;
      const verified = isoDate(
        (block as Record<string, any>)._last_verified,
      );
      const days = daysBetween(today, verified);
      if (days > 365) stale.push(`${provider}: ${days}d`);
    }
    expect(stale).toEqual([]);
  });

  test("AWS has ec2_gpu.regions.us-east-1", () => {
    const data = load();
    expect(data.aws).toHaveProperty("ec2_gpu");
    expect(data.aws.ec2_gpu).toHaveProperty("regions");
    expect(data.aws.ec2_gpu.regions).toHaveProperty("us-east-1");
  });

  test("GCP has both attached and bundled blocks", () => {
    const data = load();
    expect(data.gcp).toHaveProperty("gce_gpu_attached");
    expect(data.gcp).toHaveProperty("gce_gpu_bundled");
  });

  test("Azure has both vm_gpu and vm_vgpu blocks", () => {
    const data = load();
    expect(data.azure).toHaveProperty("vm_gpu");
    expect(data.azure).toHaveProperty("vm_vgpu");
  });

  test("serverless providers have per_gpu_second_active", () => {
    const data = load();
    for (const p of ["modal", "runpod", "replicate"]) {
      expect(data[p]).toHaveProperty("per_gpu_second_active");
    }
  });

  test("reserved providers have per_gpu_hour_reserved", () => {
    const data = load();
    for (const p of ["lambda_labs", "coreweave"]) {
      expect(data[p]).toHaveProperty("per_gpu_hour_reserved");
    }
  });

  test("every USD field is Decimal-parseable", () => {
    const data = load();
    function walk(node: unknown, path: string): void {
      if (node && typeof node === "object" && !Array.isArray(node)) {
        for (const [k, v] of Object.entries(node)) {
          if (
            typeof v === "string" &&
            (k.endsWith("_usd") ||
              k === "vcpu_count" ||
              k === "gpu_count" ||
              k === "gpu_vram_gb" ||
              k === "memory_gb")
          ) {
            try {
              new Decimal(v);
            } catch (exc) {
              throw new Error(
                `${path}.${k} not Decimal-parseable: ${v} (${String(exc)})`,
              );
            }
          } else {
            walk(v, `${path}.${k}`);
          }
        }
      } else if (Array.isArray(node)) {
        node.forEach((item, i) => walk(item, `${path}[${i}]`));
      }
    }
    walk(data, "");
  });

  test("h100-80gb-sxm5 SKU consistent across all 8 providers", () => {
    const data = load();
    const found = new Set<string>();
    function walk(node: unknown, provider: string): void {
      if (node && typeof node === "object" && !Array.isArray(node)) {
        const obj = node as Record<string, unknown>;
        if (obj.gpu_sku === "h100-80gb-sxm5") found.add(provider);
        for (const v of Object.values(obj)) walk(v, provider);
      } else if (Array.isArray(node)) {
        for (const v of node) walk(v, provider);
      }
    }
    for (const provider of Object.keys(data)) {
      if (provider === "_meta") continue;
      walk(data[provider], provider);
    }
    const expected = new Set([
      "aws",
      "gcp",
      "azure",
      "modal",
      "runpod",
      "lambda_labs",
      "coreweave",
      "replicate",
    ]);
    expect(found).toEqual(expected);
  });

  test("every dispatch billing model has a meta default", () => {
    const meta = load()._meta as Record<string, any>;
    expect(meta).toHaveProperty("default_per_instance_hour_usd");
    expect(meta).toHaveProperty("default_per_gpu_second_active_usd");
    expect(meta).toHaveProperty("default_per_gpu_hour_reserved_usd");
    expect(meta).toHaveProperty("default_per_vgpu_hour_usd");
  });

  test("sample SKU entries carry aliases arrays", () => {
    const data = load();
    const samples = [
      data.aws.ec2_gpu.regions["us-east-1"].instance_types["p5.48xlarge"],
      data.modal.per_gpu_second_active.default.h100,
      data.azure.vm_vgpu.regions.eastus.instance_types[
        "Standard_NV6ads_A10_v5"
      ],
    ];
    for (const entry of samples) {
      expect(entry).toHaveProperty("aliases");
      expect(Array.isArray(entry.aliases)).toBe(true);
    }
  });

  test("Decision #5 — GPU catalog has no arch nesting (gpu_sku flat)", () => {
    const data = load();
    const p5 =
      data.aws.ec2_gpu.regions["us-east-1"].instance_types["p5.48xlarge"];
    expect(p5).toHaveProperty("gpu_sku");
    expect(p5).not.toHaveProperty("x86_64");
    expect(p5).not.toHaveProperty("arm64");
  });

  test("_meta.version is semver-like", () => {
    const meta = load()._meta as Record<string, any>;
    const v = String(meta.version);
    const parts = v.split(".");
    expect(parts.length).toBe(3);
    for (const p of parts) {
      expect(Number.isInteger(parseInt(p, 10))).toBe(true);
    }
  });
});

File: rust/tests/gpu_catalog_integrity_test.rs
//! GPU catalog integrity tests — Phase 2 Task 4.
//!
//! Rust port of `python/tests/test_gpu_catalog_integrity.py` (commit 97f736b).
//! Pins the structural invariants of the bundled `data/gpu_prices.json` so
//! that a future refresh can't drift shape, and enforces Decision #11's
//! freshness thresholds.
//!
//! No production code — these tests load the bundled catalog and walk it.

use std::str::FromStr;

use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use serde_json::Value;

const CATALOG: &str = include_str!("../src/data/gpu_prices.json");

fn catalog() -> Value {
    serde_json::from_str(CATALOG).expect("gpu_prices.json must be valid JSON")
}

#[test]
fn catalog_has_all_eight_providers() {
    let c = catalog();
    for prov in &[
        "aws",
        "gcp",
        "azure",
        "modal",
        "runpod",
        "lambda_labs",
        "coreweave",
        "replicate",
    ] {
        assert!(
            c.get(prov).is_some(),
            "gpu_prices.json missing provider block: {}",
            prov
        );
    }
}

#[test]
fn meta_has_four_billing_model_defaults() {
    let c = catalog();
    let meta = c.get("_meta").expect("_meta block required");
    for key in &[
        "default_per_instance_hour_usd",
        "default_per_gpu_second_active_usd",
        "default_per_gpu_hour_reserved_usd",
        "default_per_vgpu_hour_usd",
    ] {
        assert!(
            meta.get(key).is_some(),
            "_meta missing default rate: {}",
            key
        );
    }
}

#[test]
fn meta_version_is_semver() {
    let c = catalog();
    let v = c
        .get("_meta")
        .and_then(|m| m.get("version"))
        .and_then(|v| v.as_str())
        .expect("_meta.version required");
    let parts: Vec<&str> = v.split('.').collect();
    assert_eq!(parts.len(), 3, "_meta.version must be X.Y.Z; got {}", v);
    for p in parts {
        p.parse::<u32>()
            .unwrap_or_else(|_| panic!("_meta.version component non-numeric: {}", v));
    }
}

#[test]
fn aws_has_ec2_gpu_regions() {
    let c = catalog();
    assert!(c["aws"]["ec2_gpu"]["regions"].is_object());
}

#[test]
fn gcp_has_both_attached_and_bundled() {
    let c = catalog();
    assert!(c["gcp"]["gce_gpu_attached"].is_object());
    assert!(c["gcp"]["gce_gpu_bundled"].is_object());
}

#[test]
fn azure_has_both_vm_gpu_and_vm_vgpu() {
    let c = catalog();
    assert!(c["azure"]["vm_gpu"].is_object());
    assert!(c["azure"]["vm_vgpu"].is_object());
}

#[test]
fn serverless_providers_have_per_gpu_second_active() {
    let c = catalog();
    for prov in &["modal", "runpod", "replicate"] {
        assert!(
            c[prov]["per_gpu_second_active"].is_object(),
            "{} must have per_gpu_second_active block",
            prov
        );
    }
}

#[test]
fn reserved_providers_have_per_gpu_hour_reserved() {
    let c = catalog();
    for prov in &["lambda_labs", "coreweave"] {
        assert!(
            c[prov]["per_gpu_hour_reserved"].is_object(),
            "{} must have per_gpu_hour_reserved block",
            prov
        );
    }
}

fn parse_iso_date(s: &str) -> Option<NaiveDate> {
    // Accept either YYYY-MM-DD or ISO-8601 datetimes.
    if let Ok(d) = NaiveDate::parse_from_str(s, "%Y-%m-%d") {
        return Some(d);
    }
    if let Ok(dt) = DateTime::parse_from_rfc3339(s) {
        return Some(dt.naive_utc().date());
    }
    None
}

#[test]
fn every_provider_has_parseable_last_verified() {
    let c = catalog();
    for prov in &[
        "aws",
        "gcp",
        "azure",
        "modal",
        "runpod",
        "lambda_labs",
        "coreweave",
        "replicate",
    ] {
        let block = &c[prov];
        let lv = block
            .get("_last_verified")
            .and_then(|v| v.as_str())
            .unwrap_or_else(|| panic!("{} missing _last_verified", prov));
        parse_iso_date(lv).unwrap_or_else(|| panic!("{} _last_verified unparseable: {}", prov, lv));
    }
}

#[test]
fn decision_11_no_provider_older_than_365_days() {
    // Hard-fail threshold per Decision #11.
    let c = catalog();
    let today = Utc::now().naive_utc().date();
    for prov in &[
        "aws",
        "gcp",
        "azure",
        "modal",
        "runpod",
        "lambda_labs",
        "coreweave",
        "replicate",
    ] {
        let lv = c[prov]
            .get("_last_verified")
            .and_then(|v| v.as_str())
            .unwrap_or("1970-01-01");
        let d = parse_iso_date(lv).unwrap_or_else(|| panic!("unparseable date: {}", lv));
        let age_days = (today - d).num_days();
        assert!(
            age_days < 365,
            "Provider {} _last_verified is {} days old (> 365); refresh required",
            prov,
            age_days
        );
    }
}

#[test]
fn all_dollar_amounts_parse_as_decimal() {
    // Tree-walk every *_usd / gpu_count / gpu_vram_gb / memory_gb string —
    // each must parse cleanly as Decimal (no float-drift surface).
    let c = catalog();
    fn walk(v: &Value, path: &str, errors: &mut Vec<String>) {
        match v {
            Value::Object(m) => {
                for (k, vv) in m {
                    let p = if path.is_empty() {
                        k.clone()
                    } else {
                        format!("{}.{}", path, k)
                    };
                    // Decimal-string-shaped keys.
                    let dec_keyed = k.ends_with("_usd")
                        || k == "vcpu_count"
                        || k == "gpu_count"
                        || k == "gpu_vram_gb"
                        || k == "memory_gb";
                    if dec_keyed {
                        if let Some(s) = vv.as_str() {
                            if Decimal::from_str(s).is_err() {
                                errors.push(format!("{} = {:?} not parseable as Decimal", p, s));
                            }
                        } else if vv.is_object() || vv.is_array() {
                            // Nested structure; recurse.
                            walk(vv, &p, errors);
                        } else if !vv.is_number() && !vv.is_null() {
                            errors.push(format!(
                                "{} has non-string non-number type {:?}",
                                p, vv
                            ));
                        }
                    } else {
                        walk(vv, &p, errors);
                    }
                }
            }
            Value::Array(a) => {
                for (i, item) in a.iter().enumerate() {
                    walk(item, &format!("{}[{}]", path, i), errors);
                }
            }
            _ => {}
        }
    }
    let mut errs = Vec::new();
    walk(&c, "", &mut errs);
    assert!(errs.is_empty(), "Decimal-shape violations:\n{}", errs.join("\n"));
}

#[test]
fn cross_provider_h100_canonical_sku_consistency() {
    // Load-bearing — `h100-80gb-sxm5` must appear across providers so a
    // customer can compare "Modal H100 vs AWS p5" through dexcost. This is
    // the catalog portability contract from §4 of the design.
    let c = catalog();
    let json_str = serde_json::to_string(&c).expect("re-serialize catalog");
    assert!(
        json_str.contains("h100-80gb-sxm5"),
        "catalog must contain canonical SKU h100-80gb-sxm5 across providers"
    );
}
